extensions/amp-image-slider/0.1/amp-image-slider.js

```diff
@@ -0,0 +1,422 @@
+  }
+
+  /**
+   * TODO
+   * @param {Event} e
+   */
+  dragStart(e) {
+    e.preventDefault();
+
+    // TODO(kqian): container or window???
+    // this.container_.addEventListener('mousemove', this.dragMove);
+    // this.container_.addEventListener('mouseup', this.dragEnd);
+    this.win.addEventListener('mousemove', this.dragMove);
+    this.win.addEventListener('mouseup', this.dragEnd);
+
+    this.moveOffset_ = e.pageX;
+    this.splitOffset_ = this.bar_.getBoundingClientRect().left;
+  }
+  /**
+   * TODO
+   * @param {Event} e
+   */
+  dragMove(e) {
+    e.preventDefault();
+    e.stopPropagation(); // avoid clashing with clickImage
+    this.pointerMoveX_(e.pageX);
+  }
+  /**
+   * TODO
+   */
+  dragEnd() {
+    // TODO(kqian): container or window???
+    // this.container_.removeEventListener('mousemove', this.dragMove);
+    // this.container_.removeEventListener('mouseup', this.dragEnd);
+    this.win.removeEventListener('mousemove', this.dragMove);
+    this.win.removeEventListener('mouseup', this.dragEnd);
+
+    this.moveOffset_ = 0;
+    this.splitOffset_ = 0;
+  }
+  /**
+   * TODO
+   * @param {Event} e
+   */
+  touchStart(e) {
+    this.container_.addEventListener('touchmove', this.touchMove);
+    this.container_.addEventListener('touchend', this.touchEnd);
+
+    this.moveOffset_ = e.touches[0].pageX;
+    this.splitOffset_ = this.bar_.getBoundingClientRect().left;
+  }
+  /**
+   * TODO
+   * @param {Event} e
+   */
+  touchMove(e) {
+    e.stopPropagation(); // avoid clashing with clickImage
+    if (e.touches.length > 0) {
+      this.pointerMoveX_(e.touches[0].pageX);
+    }
+  }
+  /**
+   * TODO
+   * @param {?Event} e
+   */
+  touchEnd(e) {
+    if (e) {
+      // Avoid bubbling up to element
+      e.stopPropagation();
+    }
+    this.container_.removeEventListener('touchmove', this.touchMove);
+    this.container_.removeEventListener('touchend', this.touchEnd);
+
+    this.moveOffset_ = 0;
+    this.splitOffset_ = 0;
+  }
+
+  /**
+   * Pointer move X logic
+   * @param {number} pointerX
+   */
+  pointerMoveX_(pointerX) {
+    const width = this.container_.offsetWidth;
+    const {left: leftBound, right: rightBound}
+        = this.container_.getBoundingClientRect();
+
+    const moveX = pointerX - this.moveOffset_;
+    const newPos = Math.max(leftBound,
+        Math.min(this.splitOffset_ + moveX, rightBound));
+    const newPercentage = (newPos - leftBound) / width;
+    this.updatePositions(newPercentage);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.scheduleLayout(this.leftAmpImage_);
+    this.scheduleLayout(this.rightAmpImage_);
+
+    Promise.all([
+      // LOAD_START
+      this.leftAmpImage_.signals().whenSignal(CommonSignals.LOAD_END),
+      this.rightAmpImage_.signals().whenSignal(CommonSignals.LOAD_END),
+    ]).then(() => this.registerEvents());
+  }
+
+  /** @override */
```

Comment:
You may want to do the same thing on `pauseCallback`. The thing to test for is whether these get correctly deregistered when you swipe away from it in `<amp-carousel type="slides">`.